File: 2015/13_Knights_of_the_Dinner_Table/Day13.clj
; Day 13: Knights of the Dinner Table
; https://adventofcode.com/2015/day/13

(ns day13
  (:require [clojure.string :as str]))

(defn permutations [coll]
  (if (empty? coll)
    '(())
    (for [x coll
          ps (permutations (disj (set coll) x))]
      (cons x ps))))

(defn parse-line [line]
  (let [[_ name gain-lose value next] 
        (re-matches #"(\w+) would (lose|gain) (\d+) happiness units by sitting next to (\w+)\." line)]
    {:name name
     :next next
     :happiness (* (Integer/parseInt value)
                  (if (= gain-lose "gain") 1 -1))}))

(defn build-happiness-map [input]
  (let [entries (map parse-line (str/split-lines input))]
    (reduce (fn [m {:keys [name next happiness]}]
             (assoc-in m [name next] happiness))
           {}
           entries)))

(defn calculate-happiness [input include-me?]
  (let [happiness-map (build-happiness-map input)
        names (set (keys happiness-map))
        n (count names)
        pairs (if include-me? (dec n) n)]
    (->> (permutations names)
         (map (fn [arrangement]
                (reduce + 
                  (for [i (range pairs)]
                    (let [current (nth arrangement i)
                          next (nth arrangement (mod (inc i) n))]
                      (+ (get-in happiness-map [current next] 0)
                         (get-in happiness-map [next current] 0)))))))
         (apply max))))

(defn part1 [input]
  (calculate-happiness input false))

(defn part2 [input]
  (calculate-happiness input true))

(let [input (slurp "Day13.txt")
      input-test (slurp "Day13_test.txt")]
  (println "Part 1:" (part1 input))
  (println "Part 1:" (part1 input-test))
  (println "Part 2:" (part2 input))
  (println "Part 2:" (part2 input-test)))

File: 2015/13_Knights_of_the_Dinner_Table/Day13.kt
// Day 13: Knights of the Dinner Table
// https://adventofcode.com/2015/day/13

fun <T> permutations(list: List<T>): List<List<T>> {
    if (list.isEmpty()) return listOf(emptyList())
    
    return list.flatMap { item ->
        val remaining = list - item
        permutations(remaining).map { subPerm -> listOf(item) + subPerm }
    }
}

fun prepareData(input: String): Map<String, Map<String, Int>> {
    val pattern = """(\w+) would (gain|lose) (\d+) happiness units by sitting next to (\w+)\.""".toRegex()
    val result = mutableMapOf<String, MutableMap<String, Int>>()
    
    input.split("\r\n")
        .mapNotNull { line -> pattern.matchEntire(line)?.groupValues }
        .forEach { (_, name, gainLose, value, next) ->
            val happiness = value.toInt() * if (gainLose == "gain") 1 else -1
            if (!result.containsKey(name)) {
                result[name] = mutableMapOf()
            }
            result[name]?.put(next, happiness)
        }
    
    return result
}

fun calculateHappiness(input: String, includeMe: Boolean): Int {
    val map = prepareData(input)
    val names = map.keys.toList()
    
    return permutations(names).maxOf { arrangement ->
        val n = arrangement.size
        val pairs = if (includeMe) n - 1 else n
        (0 until pairs).sumOf { i ->
            val current = arrangement[i]
            val next = arrangement[(i + 1) % n]
            map[current]!![next]!! + map[next]!![current]!!
        }
    }
}

fun part1(input: String) = calculateHappiness(input, false)
fun part2(input: String) = calculateHappiness(input, true)

fun main() {
    val input = java.io.File("Day13.txt").readText()
    val inputTest = java.io.File("Day13_test.txt").readText()

    println("Part 1: ${part1(input)}")
    println("Part 1: ${part1(inputTest)}")
    println("Part 2: ${part2(input)}")
    println("Part 2: ${part2(inputTest)}")
}